dragonfly/block/farmland.go

```diff
@@ -0,0 +1,100 @@
+package block
+
+import (
+	"github.com/df-mc/dragonfly/dragonfly/internal/item_internal"
+	"github.com/df-mc/dragonfly/dragonfly/item"
+	"github.com/df-mc/dragonfly/dragonfly/world"
+	"math/rand"
+)
+
+type Farmland struct {
+	// Hydration is how much moisture a block has. This is calculated by checking if there is a water source block
+	// 4 blocks away in any direction from the farmland that is either 1 block above or on the same level as the farmland block.
+	// If there isn't, we then count down the hydration level by one until it eventually dries up and turns back into dirt.
+	Hydration uint8
+}
+
+// NeighbourUpdateTick ...
+func (f Farmland) NeighbourUpdateTick(pos, block world.BlockPos, w *world.World) {
+	if _, isAir := w.Block(pos.Side(world.FaceUp)).(Air); !isAir {
+		if _, isCrop := w.Block(pos.Side(world.FaceUp)).(Crop); !isCrop {
+			w.SetBlock(pos, Dirt{})
+		}
+	}
+}
+
+// RandomTick ...
+func (f Farmland) RandomTick(pos world.BlockPos, w *world.World, r *rand.Rand) {
+	// Calculate the Hydration of the farmland block.
+	f.CalculateHydration(pos, w)
+	// Check if there is a crop on the farmland block.
+	if crop, isCrop := w.Block(pos.Add(world.BlockPos{0, 1})).(Crop); isCrop {
+		// Check if the crop requires Hydration to grow and if it meets those requirements.
+		if crop.RequiresHydration() && f.Hydration == 0 {
+			return
+		}
+
+		// Check if the crop can grow due to lighting.
+		if w.Light(pos.Add(world.BlockPos{0, 1})) >= crop.LightLevelRequired() {
+			crop.Grow(pos.Add(world.BlockPos{0, 1}), w, r, f.Hydration)
+		}
+	} else if f.Hydration <= 0 {
+		// If no crop exists and the Hydration level is 0, turn the block into dirt.
+		w.SetBlock(pos, item_internal.Dirt)
+	}
+
+	// If the farmland is dehydrated it should be replaced with it's corresponding farmland block.
+	if f.Hydration < 7 && f.Hydration > 0 {
+		// Decrease the hydration one by one down to zero.
+		w.SetBlock(pos, Farmland{f.Hydration - 1})
+	}
+}
+
+// CalculateHydration determines the Hydration or moisture of a block by scanning a 4 block box
+// around the Farmland block in each direction for water source blocks.
+// This also takes into account water source blocks one block higher than the farmland.
+func (f Farmland) CalculateHydration(pos world.BlockPos, w *world.World) {
```

Comment:
Might be easier to read this code if you changed this function to something like 'hydrated() bool` and setting hydrated farmland if it is, or hydration-1 if not. Currently it seems to do two things at once and it seems to make the code in RandomTick a bit awkward.

Reply:
I decrease the hydration after using the current hydration to continue to have more effectiveness within crop growth

Reply:
Problem is that at the moment you are forced to set the hydration to 6 here if the block is not hydrated, but then in RandomTick you end up dropping another level. I don't think that's how it's supposed to work in vanilla.

Reply:
Alright, that makes more sense